### ISSUE TITLE
[Alpine]Test failed: JIT_superpmi._superpmicollect_superpmicollect_._superpmicollect_superpmicollect_sh

### ISSUE THREAD

**Issue by VincentBu** [dotnet/coreclr#25273]
**Job:**
https://mc.dot.net/#/user/coreclr-outerloop-jitstress/ci~2Fdotnet~2Fcoreclr~2Frefs~2Fheads~2Fmaster/test~2Ffunctional~2Fcli~2F/20190618.1/workItem/JIT/analysis/xunit/JIT_superpmi._superpmicollect_superpmicollect_~2F_superpmicollect_superpmicollect_sh

**Failed tests:**
JIT_superpmi._superpmicollect_superpmicollect_._superpmicollect_superpmicollect_sh

**Log:**
```
Stack overflow.
/root/helix/work/workitem/JIT/superpmi/superpmicollect/superpmicollect.sh: line 266:    85 Aborted                 (core dumped) $LAUNCHER $ExePath \"${CLRTestExecutionArguments[@]}\"

Return code:      1
Raw output file:      /root/helix/work/workitem/JIT/superpmi/Reports/JIT.superpmi/superpmicollect/superpmicollect.output.txt
Raw output:
BEGIN EXECUTION
/root/helix/work/correlation/corerun superpmicollect.exe ''
SuperPMI collection and playback - BEGIN
Setting environment variables:
    SuperPMIShimLogPath=/tmp/wnwrrstv.4i2SPMI
    SuperPMIShimPath=/root/helix/work/correlation/libclrjit.so
    COMPlus_AltJit=*
    COMPlus_AltJitName=libsuperpmi-shim-collector.so
Running: /usr/bin/perl -pi -e 's/\\r\
|\
|\\r/\
/g' \"/root/helix/work/workitem/JIT/superpmi/superpmicollect/Bytemark.sh\"
Expected: 100
Actual: 134
END EXECUTION - FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=/root/helix/work/correlation
> /root/helix/work/workitem/JIT/superpmi/superpmicollect/superpmicollect.sh
Expected: True
Actual:   False
```

**Comment by RussKeldorph** [comment]
Stack overflow on Alpine/ARM64.

**Comment by echesakovMSFT** [comment]
Also fail on **Alpine.38.Amd64.Open-x64:Checked-zapdisable** https://mc.dot.net/#/user/coreclr-outerloop-jitstress/ci~2Fdotnet~2Fcoreclr~2Frefs~2Fheads~2Fmaster/test~2Ffunctional~2Fcli~2F/20190623.1/workItem/JIT/wilogs
```
export COMPlus_TieredCompilation=0
export COMPlus_ReadyToRun=0
export COMPlus_ZapDisable=1

  Starting:    JIT.superpmi.XUnitWrapper
    JIT/superpmi/superpmicollect/superpmicollect.sh [FAIL]
      Stack overflow.
      /root/helix/work/workitem/JIT/superpmi/superpmicollect/superpmicollect.sh: line 266:  4296 Aborted                 (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
      
      Return code:      1
      Raw output file:      /root/helix/work/workitem/JIT/superpmi/Reports/JIT.superpmi/superpmicollect/superpmicollect.output.txt
      Raw output:
      BEGIN EXECUTION
      /root/helix/work/correlation/corerun superpmicollect.exe ''
      SuperPMI collection and playback - BEGIN
      Setting environment variables:
          SuperPMIShimLogPath=/tmp/1kfgt2xh.jlxSPMI
          SuperPMIShimPath=/root/helix/work/correlation/libclrjit.so
          COMPlus_AltJit=*
          COMPlus_AltJitName=libsuperpmi-shim-collector.so
      Running: /usr/bin/perl -pi -e 's/\r\n|\n|\r/\n/g' "/root/helix/work/workitem/JIT/superpmi/superpmicollect/Bytemark.sh"
      Expected: 100
      Actual: 134
      END EXECUTION - FAILED
      Test Harness Exitcode is : 1
      To run the test:
      > set CORE_ROOT=/root/helix/work/correlation
      > /root/helix/work/workitem/JIT/superpmi/superpmicollect/superpmicollect.sh
      Expected: True
      Actual:   False
      Stack Trace:
        /__w/8/s/bin/tests/Linux.x64.Checked/TestWrappers/JIT.superpmi/JIT.superpmi.XUnitWrapper.cs(129,0): at JIT_superpmi._superpmicollect_superpmicollect_._superpmicollect_superpmicollect_sh()
      Output:
        Stack overflow.
        /root/helix/work/workitem/JIT/superpmi/superpmicollect/superpmicollect.sh: line 266:  4296 Aborted                 (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
        
        Return code:      1
        Raw output file:      /root/helix/work/workitem/JIT/superpmi/Reports/JIT.superpmi/superpmicollect/superpmicollect.output.txt
        Raw output:
        BEGIN EXECUTION
        /root/helix/work/correlation/corerun superpmicollect.exe ''
        SuperPMI collection and playback - BEGIN
        Setting environment variables:
            SuperPMIShimLogPath=/tmp/1kfgt2xh.jlxSPMI
            SuperPMIShimPath=/root/helix/work/correlation/libclrjit.so
            COMPlus_AltJit=*
            COMPlus_AltJitName=libsuperpmi-shim-collector.so
        Running: /usr/bin/perl -pi -e 's/\r\n|\n|\r/\n/g' "/root/helix/work/workitem/JIT/superpmi/superpmicollect/Bytemark.sh"
        Expected: 100
        Actual: 134
        END EXECUTION - FAILED
        Test Harness Exitcode is : 1
        To run the test:
        > set CORE_ROOT=/root/helix/work/correlation
        > /root/helix/work/workitem/JIT/superpmi/superpmicollect/superpmicollect.sh
  Finished:    JIT.superpmi.XUnitWrapper
```

**Comment by BruceForstall** [comment]
I repro'ed this by building (debug flavor) on Linux/x64 with Docker image `mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.6-WithNode-cfdd435-20190521001804`, then running under Docker image `mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.8-helix-45b1fa2-20190327215821`

```
bash-4.4$ /home/bruce/gh/coreclr2/bin/tests/Linux.x64.Debug/Tests/Core_Root/corerun ./superpmicollect.exe
SuperPMI collection and playback - BEGIN
Setting environment variables:
    SuperPMIShimLogPath=/tmp/tyhhipre.quaSPMI
    SuperPMIShimPath=/home/bruce/gh/coreclr2/bin/tests/Linux.x64.Debug/Tests/Core_Root/libclrjit.so
    COMPlus_AltJit=*
    COMPlus_AltJitName=libsuperpmi-shim-collector.so
Running: /usr/bin/perl -pi -e 's/\r\n|\n|\r/\n/g' "/home/bruce/gh/coreclr2/bin/tests/Linux.x64.Debug/JIT/superpmi/superpmicollect/Bytemark.sh"
Can't remove /home/bruce/gh/coreclr2/bin/tests/Linux.x64.Debug/JIT/superpmi/superpmicollect/Bytemark.sh: Permission denied, skipping file.
Stack overflow.
Aborted (core dumped)
```

Can't figure out how to get a core dump written:
```
bash-4.4$ cat /proc/sys/kernel/core_pattern
|/usr/share/apport/apport %p %s %c %d %P

bash-4.4$ echo '/tmp/core_%e.%p' | sudo tee /proc/sys/kernel/core_pattern
tee: /proc/sys/kernel/core_pattern: Read-only file system
/tmp/core_%e.%p
```

**Comment by BruceForstall** [comment]
Also can't figure out how to install a debugger

**Comment by BruceForstall** [comment]
@janvorli Can you offer any suggestions on debugging in an Alpine container?

**Comment by BruceForstall** [comment]
Also, @jashook, in case you can offer help here.

**Comment by BruceForstall** [comment]
@jashook helped me offline get gdb installed and working in the Docker container.

In the test case, with ReadyToRun disabled, the JIT is in a deep morph recursion -- about 30 frames. I calculated about 85K of stack usage. It doesn't seem completely out of the ordinary. I would expect to see this happen frequently, in fact.

@janvorli: how much stack do we get on Alpine? (and how can I figure that out)

**Comment by janvorli** [comment]
@BruceForstall secondary threads created by CoreCLR PAL on MUSL get 1.5MB of stack by default (PAL overrides the default provided by MUSL, which is around 80kB). Threads created directly by pthread_create or C++ thread class get the MUSL provided default stack size. 
Primary thread's stack size reflects the `ulimit -s` size.
You can check the size of the stack by dumping memory map using "info proc maps" GDB command and searching for the range containing the SP at the crash time. The size and base address of the mapping would show the size of the mapped portion of the stack at the crash time.

**Comment by BruceForstall** [comment]
I'm having trouble debugging this, and understanding what gdb is telling me. I have a reliable repro. I've been using:
```
COMPlus_ReadyToRun=0
COMPlus_JitDump=TZif_ParseRaw
COMPlus_TieredCompilation=0
```
I get:
```
Thread 9 "corerun" received signal SIGSEGV, Segmentation fault.
[Switching to LWP 339]
0x00007ffff48eec9c in Compiler::fgMorphSmpOp (this=0x5555558d2048, tree=0x5555559cef38, mac=0x0) at /home/bruce/gh/coreclr2/src/jit/morph.cpp:11295
11295       ALLOCA_CHECK();
(gdb) disas
Dump of assembler code for function Compiler::fgMorphSmpOp(GenTree*, Compiler::MorphAddrContext*):
   0x00007ffff48eec60 <+0>:     push   rbp
   0x00007ffff48eec61 <+1>:     mov    rbp,rsp
   0x00007ffff48eec64 <+4>:     sub    rsp,0x1570
   0x00007ffff48eec6b <+11>:    mov    rax,QWORD PTR fs:0x28
   0x00007ffff48eec74 <+20>:    mov    QWORD PTR [rbp-0x8],rax
   0x00007ffff48eec78 <+24>:    mov    QWORD PTR [rbp-0x108],rdi
   0x00007ffff48eec7f <+31>:    mov    QWORD PTR [rbp-0x78],rsi
   0x00007ffff48eec83 <+35>:    mov    QWORD PTR [rbp-0x110],rdx
   0x00007ffff48eec8a <+42>:    mov    rax,QWORD PTR [rbp-0x108]
   0x00007ffff48eec91 <+49>:    lea    rdi,[rbp-0x80]
   0x00007ffff48eec95 <+53>:    mov    QWORD PTR [rbp-0x3f0],rax
=> 0x00007ffff48eec9c <+60>:    call   0x7ffff490a0a0 <AllocaCheck::AllocaCheck()>
```

I presume the actual SIGSEGV is at the `mov    QWORD PTR [rbp-0x3f0],rax` -- right? The call targets code that does exist.

In this case, we have:
```
rbp            0x7ffff1e51e80
rsp            0x7ffff1e50910
```

```
(gdb) info proc mappings
...
          Start Addr           End Addr       Size     Offset objfile
      0x7fff7e079000     0x7fffee712000 0x70699000        0x0
      0x7ffff1e4b000     0x7ffff1e4c000     0x1000        0x0
      0x7ffff1e4c000     0x7ffff1e4f000     0x3000        0x0
      0x7ffff1e4f000     0x7ffff1e51000     0x2000        0x0 <--- looks like rsp is in here
      0x7ffff1e51000     0x7ffff1e66000    0x15000        0x0
...

Unfortunately, this doesn't show the type of section (mapped, guard, etc.). It also looks like there are only 2 pages here where SP lies.

I thought maybe the native compiler wasn't doing enough stack probing, since it allocates 0x1570 bytes without a specific probe. But the code above, `mov    rax,QWORD PTR [rbp-0x108]` didn't crash, and that would be on the same page as `mov    QWORD PTR [rbp-0x3f0],rax`, which apparently did SEGV.

ulimit -s shows "8192" which is much bigger than what I'd expect (8MB, versus 85kB?). This is running on thread 9, but I can't tell if that's a PAL created secondary thread or the MUSL default stack size thread. I presume a PAL thread, due it it being 9 and not 1, but maybe that's not how that works.

Unfortunately, stack walking doesn't seem to work through the managed frames, so we don't get to the real base function:
```
#53 0x00007ffff6c00d74 in ThePreStub () at /home/bruce/gh/coreclr2/src/vm/amd64/theprestubamd64.S:17
#54 0x00007fff7dee25f8 in ?? ()
#55 0x00007ffff1e65338 in ?? ()
#56 0x00007ffff1e65330 in ?? ()
#57 0x00007ffff1e65328 in ?? ()
#58 0x0000000000000000 in ?? () <==== why 0x0?
```

And I guess we can't run LLDB + SOS in an Alpine container?

@janvorli any suggestions?

**Comment by BruceForstall** [comment]
The problem happens is in the JIT compilation of `Process.WaitForExit()`. I reduced the test case to:

```
using System;
using System.IO;
using System.Diagnostics;

namespace SpawnProcess
{
    internal class Program
    {
        private static int Main(string[] args)
        {
            int result = 100;

            try
            {
                string program = "/bin/ls";
                string arguments = "";
                Console.WriteLine("Running: " + program + " " + arguments);
                Process p = Process.Start(program, arguments);
                p.WaitForExit();
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine("ERROR: unknown exception: " + ex.Message);
                result = 101;
            }

            return result;
        }
    }
}
```

In particular,
```
m_pszDebugMethodName = 0x7fff7dad0d90 "FindEntry",
m_pszDebugClassName = 0x7fff7dad05c0 "System.Collections.Generic.Dictionary`2[Int32,__Canon]",
m_pszDebugMethodSignature = 0x7fff7dad25d0 "instance int32 *(!0)",
```

@janvorli Can you try to repro using this test case, and offer suggestions on what the root cause of the problem is? In particular, are we actually using 1.5MB of stack?

One thing I note is the bottom of the stack is:
```
#51 0x00007ffff6c17d04 in ThePreStub () at /home/bruce/gh/coreclr2/src/vm/amd64/theprestubamd64.S:17
#52 0x00007fff7d664747 in ?? ()
#53 0x0000000000000000 in ?? ()
```

so do we have a problem doing stack walks (in gdb) through managed code?

**Comment by janvorli** [comment]
> so do we have a problem doing stack walks (in gdb) through managed code?

GDB cannot walk through managed frames on any Unix platform. It can unwind only frames that have DWARF unwind info. LLDB is more clever and has a fallback mechanism where it tries to use RBP chain when there is no unwind info, so it can unwind managed frames (although sometimes during EH, it misses part of the stack).

> ulimit -s shows "8192" which is much bigger than what I'd expect (8MB, versus 85kB?).

This limit influences only the main thread, not the secondary ones.

As for the info from mappings, it is obvious that the stack is only 0x15000 bytes large. There are two guard pages (the mapping at 0x7ffff1e4f000) that the SP is in. 

So it would seem like a thread that was not created by PAL. But it is not clear to me how that could happen in your simple case. Let me try to repro this.

**Comment by janvorli** [comment]
Ok, this is a dup of an issue I've fixed in https://github.com/dotnet/corefx/pull/38743. The thread we crash on is created by SystemNative_InitializeTerminalAndSignalHandling (in System.Native.so from the corefx repo).
The problem is in how the issue was fixed. I've originally made the stack size setting unconditional for all configurations. However in the discussion in the PR, @stephentoub has suggested to make the change debug build only, as the stack overflow issue really happens only in debug builds (due to the C++ compiler generating enormous frames for some of the JIT methods in non-optimized builds). So I've made that change and merged it in that way.
What none of us has realized though is that when we run debug / checked build of coreclr tests, we still use release build of corefx. So the fix is not compiled in. And the debug / checked build of coreclr JIT uses the large frames and so it hits the stack overflow.

**Comment by BruceForstall** [comment]
Interesting. Given this, would it be appropriate to change the fix in https://github.com/dotnet/corefx/pull/38743 so it is all-configuration, not just DEBUG?

**Comment by janvorli** [comment]
We can either do that or, as @sandreenko has suggested offline, refactor the fgMorphSmpOp so that the compiler doesn't create 5kB large frame for it is in the debug build. He felt like that the function would be worth refactoring.

**Comment by BruceForstall** [comment]
TL;DR: there is a bug in the Linux build process whereby Checked builds are not being optimized.

Full info:

Assuming fgMorphSmpOp is the main culprit (and it does recurse, and appear on the stack many times in this test), I took a look at its stack size:

- Windows:
  - Release: 0x110
  - Checked: 0x170
  - Debug: 0x810
- Alpine:
  - Release: 0x138
  - Checked: 0x1570
  - Debug: 0x1570

So the interesting points:

- This function is large, with many cases, and benefits greatly from the stack packing optimization (so variables in non-overlapping scopes overlap the space they use in the frame), or perhaps other optimizations that reduce stack size.
- VC++ is a bit better than clang at packing the stack than clang, even in Debug. Maybe clang generates a lot more temps.
- Checked and Debug have the same, large stack size.

This last point surprised me. I looked into the actual compilation flags, and for Alpine we are compiling with default optimization (no "-O" flag) for Debug and Checked. Release passes "-O3". It turns out this is because configureoptimization.cmake, which sets the compilation options, depends, for Linux, on the `CLR_CMAKE_PLATFORM_UNIX` symbol being defined. But it is included in configurecompiler.cmake _before_ `CLR_CMAKE_PLATFORM_UNIX` is defined. Moving the include lower leads to Checked being built with -O2, and then I see:

- Alpine:
  - Checked: 0x198

(Debug and Release don't change.)

And, the tests pass. Since we run almost all CI testing with Checked builds, this should be a good fix.

What's interesting is all our Checked Linux testing should get faster. Hopefully we don't see any clang/llvm optimization bugs as a result of enabling the optimizer in Checked.

I submitted a PR with the fix in https://github.com/dotnet/coreclr/pull/25801
